File: Challenge3/puzzle3.rb
require_relative './claim.rb'

class Puzzle3
  attr_accessor :claims, :fabric, :overlap

  def initialize
    @fabric = generate_fabric
    @overlap = 0
    @claims = []
  end

  def generate_fabric
    arr = []
    i = 0;
    while i < 1100
      j = 0
      new_arr = []
      while j < 1100
        new_arr << 0
        j += 1
      end
      arr << new_arr
      i += 1
    end
    arr
  end

  def input_claims
    File.open('./input.txt').each do |line|
      parse_line(line)
    end
    find_all_overlap
    puts overlap
  end

  def parse_line(line)
    claim = Claim.new(line.split(' '))
    claims << claim
    add_to_fabric(claim)
  end

  def add_to_fabric(claim)
    (1..claim.width).each do |i|
      (1..claim.height).each do |j|
        row = claim.vertical_offset + j
        column = claim.horizontal_offset + i
        fabric[row][column] = fabric[row][column].zero? ? 1 : 2
      end
    end
  end

  def find_all_overlap
    fabric.each do |row|
      @overlap += row.select { |inch| inch == 2 }.count
    end
  end
end

puz = Puzzle3.new
puz.input_claims

File: Challenge3/Puzzle3.java
import java.util.*;
import java.io.*;

public class Puzzle3 {
  private ArrayList<int[]> visitedCoords = new ArrayList<int[]>();
  private ArrayList<Claim> claims = new ArrayList<Claim>();
  public final static int HEIGHT = 1100;
  public final static int WIDTH = 1100;
  private int[][] fabric = new int[Puzzle3.HEIGHT][Puzzle3.WIDTH];
  private int overlap = 0;

  public int getOverlap() {
    return this.overlap;
  }

  public void findAllOverlap() {
    parseInput();
    addClaimsToFabric();
    findOverlap();
  }

  private void parseInput() {
    try {
      FileInputStream fstream = new FileInputStream("input.txt");
      BufferedReader reader = new BufferedReader(new InputStreamReader(fstream));
      String line;
      while((line = reader.readLine()) != null) {
        parseLine(line);
      }
    } catch (FileNotFoundException e) {
      System.err.println("Error: " + e.getMessage());
    } catch (IOException e) {
      System.err.println("Error: " + e.getMessage());
    }
  }

  private void parseLine(String line) {
    Claim claim = new Claim(line);
    claims.add(claim);
  }

  private void addClaimsToFabric() {
    for(Claim claim : claims) {
      for(int i = 1; i <= claim.getHeight(); i++) {
        int row = claim.getVerticalOffset() + i;
        
        for(int j = 1; j <= claim.getWidth(); j++) {
          int column = claim.getHorizontalOffset() + j;
          if(fabric[row][column] == 0) {
            fabric[row][column] = 1;
          } else {
            if(fabric[row][column] == 1) {
              int[] coords = { row, column };
              visitedCoords.add(coords);
            }
            fabric[row][column] = 2;
          }
        }
      }
    }
  }

  private void findOverlap() {
    this.overlap = visitedCoords.size();
  }

  public String findNoOverlap() {
    String claimId = "";
    for(Claim claim : claims) {
      boolean foundClaim = true;
      for(int i = 1; i <= claim.getHeight(); i++) {
        int row = claim.getVerticalOffset() + i;

        for(int j = 1; j <= claim.getWidth(); j++) {
          int column = claim.getHorizontalOffset() + j;
          if(fabric[row][column] == 2) {
            foundClaim = false;
            break;
          }
        }

        if(foundClaim == false) { break; }
      }

      if(foundClaim == true) {
        claimId = claim.getClaimNumber();
        break;
      }
    }
    // will never hit this
    return claimId;
  }

  public static void main(String[] args) {
    Puzzle3 puz = new Puzzle3(); 
    puz.findAllOverlap();
    System.out.println(puz.getOverlap());
    String idNoOverlap = puz.findNoOverlap();
    System.out.println(idNoOverlap);
  } 
}
